src/dmd/cppmangle.d

```diff
@@ -995,8 +995,19 @@ private final class CppMangleVisitor : Visitor
         // fake mangling for fields to fix https://issues.dlang.org/show_bug.cgi?id=16525
         if (!(d.storage_class & (STC.extern_ | STC.field | STC.gshared)))
         {
-            d.error("Internal Compiler Error: C++ static non-`__gshared` non-`extern` variables not supported");
-            fatal();
+            if (!target.cpp.threadLocalSupport())
+            {
+                // OSX emits TLS wrapper functions, shouldn't be hard to do.
+                // These start with _ZTW
+                d.error("C++ `thread_local` is unsupported for this target");
+                fatal();
```

Comment:
Nope, *all* C++ compilers emit a TLS wrapper function which need to be called when interfacing with a variable such as:
```
extern(C++) extern int tlsVar;
```
So the only TLS variables that are safe to be `extern(C++)` are those that are defined on the D side - however there may be linker/runtime issues on the C++ if they try to access them (which is probably what you found out on OSX/clang, but this is conveniently hidden on GCC thanks to the kind of variables you're using to test not requiring any special construction.

Until we do this, we can't claim to have TLS support for C++ variables at all.

Reply:
For example:
```C++
#include <assert.h>
struct S
{
    S(int f);
    int field;
};
S::S(int f) : field(f) { }

extern thread_local S tls;

void fromCxx()
{
    assert(tls.field == 0xfeebdaed);
}
```
```D
extern(C++)
{
    struct S
    {
        this(int f);
        int field;
    }
    S tls;
    void fromCxx();
}

void main()
{
    tls = S(0xdeadbeef);
    fromCxx();
}

```
This won't link at all.

Reply:
To flip the test around for variables defined on the C++ side.
```C++
#include <assert.h>
struct S
{
    S(unsigned f);
    unsigned field;
};
S::S(unsigned f) : field(f) { }

thread_local S tls = S(0xdeadbeef);

void fromCxx()
{
    assert(tls.field == 0xdeadbeef);
}
```
```D
extern(C++)
{
    struct S
    {
        this(uint f);
        uint field;
    }
    extern S tls;
    void fromCxx();
}

void main()
{
    assert(tls.field == 0xdeadbeef); // FAILS.
    fromCxx();
    assert(tls.field == 0xdeadbeef); // PASSES.
}
```

Reply:
added that test case. lets see what fails